### PR TITLE
ARC-1405 Update JWT validation mechanism

### PR DESCRIPTION
**Pull request by Harminder84**
**What's in this PR?**
We relied on `jiraHost` passed as a request parameter and validation of JWT using the `sharedSecret`. But as per documentation, it states :
- Parse JWT
- Extract issuer which equals to `clientKey`
- query the installation table using client key and fetch the `jirahost`

Updated the middleware to implemented the above guidelines and set the jiraHost in `res.locals`. Also, removed the `jiraHost` cookie altogether. 

**Why**
To fix the vulnerability 

**Added feature flags**
NEW_JWT_VALIDATION

**Affected issues**  
[ARC-1405]

**How has this been tested?**  
Unit Testing 
Local

[ARC-1405]: https://softwareteams.atlassian.net/browse/ARC-1405

### CODE REVIEW

File: src/middleware/jira-jwt-middleware.ts
Comment (gxueatlassian) [inline src/middleware/jira-jwt-middleware.ts:22]: Hmm, this just return but not calling `next?`

File: src/middleware/jira-jwt-middleware.ts
Comment (gxueatlassian) [inline src/middleware/jira-jwt-middleware.ts:30]: why this line `req.session.jiraHost = installation.jiraHost;`?

File: src/middleware/jira-jwt-middleware.ts
Comment (gxueatlassian) [inline src/middleware/jira-jwt-middleware.ts:53]: same `return` without calling `next`?

File: src/middleware/jira-jwt-middleware.ts
Comment (gxueatlassian) [inline src/middleware/jira-jwt-middleware.ts:47]: I am not sure `res.locals.jwtVerified`, as the convention is anything within `res.locals` is "verified". 
let's discuss a bit ?

File: src/middleware/jira-jwt-middleware.ts
Comment (Harminder84) [reply src/middleware/jira-jwt-middleware.ts:22]: `getInstallationAndTokenVerification` method sets proper msg on response object, so just return in case of any error.

File: src/middleware/jira-jwt-middleware.ts
Comment (Harminder84) [reply src/middleware/jira-jwt-middleware.ts:30]: This is a case when application accessed directly and jwt token migrated as a cookie. In that case jirahost to be kept in session and remove jwt cookie.

File: src/middleware/jira-jwt-middleware.ts
Comment (gxueatlassian) [reply src/middleware/jira-jwt-middleware.ts:22]: Since `getInstallationAndTokenVerification` is completely private to this file, can you refactor it a bit so that the return is more explicity (like we don't implicitly depends on it to call `res.status().send()`)

It could be error prone implicitly depends on it to do something. I'd prefer having it return more explicit result either success or fail with error code (or throw with error )

File: src/middleware/jira-jwt-middleware.ts
Comment (Harminder84) [reply src/middleware/jira-jwt-middleware.ts:47]: This is for performance improvement. If token is already verified in `jiraHostMiddleware`, don't verify it again.

File: src/middleware/jira-jwt-middleware.ts
Comment (gxueatlassian) [inline src/middleware/jira-jwt-middleware.ts:11]: /nit, this can be 
```
if (!token) {
  res.locals.jiraHost = req.session.jiraHost;
  return next();
}
```

**Review by gxueatlassian — CHANGES_REQUESTED** [review]
Just blocking it for now, let's discuss tmr.

**Review by gxueatlassian — COMMENTED** [review]
just unblock it.

File: src/middleware/jiraSymmetricJwtMiddleware.ts
Comment (bgvozdev) [inline src/middleware/jiraSymmetricJwtMiddleware.ts:73]: Should we maybe log the error, too?
```
catch (err) {
  logger.warn({ err },  whatever);
  throw ...
}
```

File: src/middleware/frontend-log-middleware.ts
Comment (bgvozdev) [inline src/middleware/frontend-log-middleware.ts:75]: That's for logging only, right?

I think we could change it to something like
```
const createJiraHostChildLogger = (logger: Logger) => {
    ....

}
```

and encapsulate all the related to jiraHost logic here, removing all confusing words like "validated/unsafe/validity" etc. Wdyt?

File: src/middleware/jiraSymmetricJwtMiddleware.test.ts
Comment (bgvozdev) [inline src/middleware/jiraSymmetricJwtMiddleware.test.ts:39]: Instead of mocking everything, wouldn't it be easer to create a record in Db? E.g. like we do here:

http://github.com/atlassian/github-for-jira/blob/1fc1bfb87eb30b4a805aa9990461b203d4825575/src/sqs/push.test.ts#L109-L109

File: src/middleware/jiraSymmetricJwtMiddleware.ts
Comment (bgvozdev) [inline src/middleware/jiraSymmetricJwtMiddleware.ts:20]: I feel like this should be a warn

File: src/middleware/jiraSymmetricJwtMiddleware.ts
Comment (bgvozdev) [inline src/middleware/jiraSymmetricJwtMiddleware.ts:26]: Same: I feel like this should be a warn

File: src/middleware/jiraSymmetricJwtMiddleware.ts
Comment (bgvozdev) [inline src/middleware/jiraSymmetricJwtMiddleware.ts:37]: Same: I feel like this should be a warn. Also, would it make sense to add error as the first arg?
```
catch (err) {
req.log.warn({ err }, "Could not verify symmetric JWT");
}
```

File: src/middleware/jiraSymmetricJwtMiddleware.ts
Comment (bgvozdev) [inline src/middleware/jiraSymmetricJwtMiddleware.ts:32]: shouldn't it be `req.originalURL` ? https://stackoverflow.com/a/68372568

File: src/middleware/jiraSymmetricJwtMiddleware.test.ts
Comment (bgvozdev) [inline src/middleware/jiraSymmetricJwtMiddleware.test.ts:20]: also, instead of mocking everything, would it make sense to create the app? This way we could test `req.path` stuff, too, couldn't we? (We might make a mistake during mocking..)  Like here:
http://github.com/atlassian/github-for-jira/blob/1fc1bfb87eb30b4a805aa9990461b203d4825575/src/routes/api/api-router.test.ts#L19-L19

File: src/middleware/jiraSymmetricJwtMiddleware.ts
Comment (bgvozdev) [inline src/middleware/jiraSymmetricJwtMiddleware.ts:54]: should we log a error?

Comment (bgvozdev) [reply src/middleware/jiraSymmetricJwtMiddleware.ts:54]: also, probably we should also return res.status(401) ?

File: src/middleware/jiraSymmetricJwtMiddleware.ts
Comment (bgvozdev) [inline src/middleware/jiraSymmetricJwtMiddleware.ts:87]: should we log the error here, too?

File: src/middleware/jiraSymmetricJwtMiddleware.ts
Comment (bgvozdev) [inline src/middleware/jiraSymmetricJwtMiddleware.ts:100]: should we log the error here, too?

File: src/middleware/jiraSymmetricJwtMiddleware.ts
Comment (bgvozdev) [inline src/middleware/jiraSymmetricJwtMiddleware.ts:1]: I think the name of the file is not following the convention, should be
`jira-symmetric-jwt-middleware.ts`

File: src/middleware/frontend-log-middleware.ts
Comment (Harminder84) [reply src/middleware/frontend-log-middleware.ts]: This file doesn't require any change. Changes reverted.

File: src/middleware/jira-symmetric-jwt-middleware.ts
Comment (Harminder84) [reply src/middleware/jira-symmetric-jwt-middleware.ts:16]: Do you know any scenario where we're passing  JWT in header?

File: src/middleware/jira-symmetric-jwt-middleware.ts
Comment (Harminder84) [reply src/middleware/jira-symmetric-jwt-middleware.ts:18]: I don't prefer shorter route cuz of readability but we can break into sub methods if it improves readability

File: src/middleware/jira-symmetric-jwt-middleware.ts
Comment (Harminder84) [reply src/middleware/jira-symmetric-jwt-middleware.ts:33]: This middleware only deals with Symmertric tokens not Asymmertric, so there is no need to determine token type.
Now Symmetric token could be with fixed or computed qsh.  That is decided using `checkPathValidity` method.

File: src/middleware/jira-symmetric-jwt-middleware.ts
Comment (gxueatlassian) [reply src/middleware/jira-symmetric-jwt-middleware.ts:16]: Nah, just a thought. But usually token thing is safer in the headers, which we should be practicing more. As all middlewares (switch, gateway) will possibly logs the url while ignore logging of headers especially if they are in `Authorization` header.

Comment (gxueatlassian) [reply src/middleware/jira-symmetric-jwt-middleware.ts:16]: All good. Can close this.

File: src/middleware/jira-symmetric-jwt-middleware.ts
Comment (gxueatlassian) [reply src/middleware/jira-symmetric-jwt-middleware.ts:18]: Okay, I can leave this to you. All good.

File: src/middleware/jira-symmetric-jwt-middleware.ts
Comment (gxueatlassian) [reply src/middleware/jira-symmetric-jwt-middleware.ts:33]: `Symmetric token could be with fixed or computed qsh` <-- this is the `TokenType` I meant, `normal` or `context`

So I think the determine it is `fixed` or `computed qsh`, we should just look into the token itself, instead of depends on url.

---

However, to determine whether `computed qsh` or `fixed` is allowed, we should look into url. This is a separate thing compare to above.

----

But all goods, since this is the existing logics, we can leave this as it is.

Comment (gxueatlassian) [reply src/middleware/jira-symmetric-jwt-middleware.ts:33]: /nit
maybe rename `tokenType` to `expectedTokenType` or `allowedTokenType`? So that we know this is NOT the type of the token itself, but the type we are expecting (from url?)

File: src/middleware/jira-symmetric-jwt-middleware.ts
Comment (gxueatlassian) [inline src/middleware/jira-symmetric-jwt-middleware.ts:63]: 1. Should it be `warn`? 
2. Change text to `No token found and session cookie has not jiraHost`?

File: src/middleware/jira-symmetric-jwt-middleware.ts
Comment (gxueatlassian) [inline src/middleware/jira-symmetric-jwt-middleware.ts:19]: Should we put a try/catch? If `getIssuer` throw errors, what's the expected output?

File: src/routes/github/create-branch/github-create-branch-options-get.ts
Comment (gxueatlassian) [inline src/routes/github/create-branch/github-create-branch-options-get.ts:27]: **IMPORTANT**

This is concerning. Need more details on this.

File: src/routes/jira/connect/enterprise/app/jira-connect-enterprise-app-router.ts
Comment (gxueatlassian) [inline src/routes/jira/connect/enterprise/app/jira-connect-enterprise-app-router.ts:18]: **IMPORTANT**
This is missing the new middleware (for `res.locals.jiraHost`)

Comment (gxueatlassian) [reply src/routes/jira/connect/enterprise/app/jira-connect-enterprise-app-router.ts:18]: Tbh, I think we need another refactoring (not this PR).
I spend quite a bit of time and look carefully to find this bug. Feels like:
1. We either missing tests (which is the case, but hard to fix)
2. And this is example of why nesting router doesn't help code readability from security point of view.

**Review by gxueatlassian — CHANGES_REQUESTED** [review]
Some comments remaining.
And maybe we can have another walkthrough?

File: src/middleware/jira-symmetric-jwt-middleware.ts
Comment (Harminder84) [reply src/middleware/jira-symmetric-jwt-middleware.ts]: Refactored the code, it will return 401 if getIssuer `throws` error

File: src/middleware/jira-symmetric-jwt-middleware.ts
Comment (Harminder84) [reply src/middleware/jira-symmetric-jwt-middleware.ts:36]: I think this will create ambigituy, leaving as it is for now.

File: src/routes/jira/connect/enterprise/app/jira-connect-enterprise-app-router.ts
Comment (Harminder84) [reply src/routes/jira/connect/enterprise/app/jira-connect-enterprise-app-router.ts]: Added `jiraSymmetricJwtMiddleware` for `JiraConnectEnterpriseAppRouter` router

File: src/routes/github/github-oauth-router.ts
Comment (Harminder84) [inline src/routes/github/github-oauth-router.ts:23]: Since `GithubAuthMiddleware` is now bind with route, so `baseUrl` contains full path. 
For e.g. for route `/github/configuration`, earlier it `baseUrl` contains `/github`, now it has `/github/configuration`

File: src/middleware/jira-symmetric-jwt-middleware.test.ts
Comment (bgvozdev) [inline src/middleware/jira-symmetric-jwt-middleware.test.ts:129]: nit: that already was tested above (in `should return valid response when jiraHost set in session`)

File: src/middleware/jira-symmetric-jwt-middleware.test.ts
Comment (bgvozdev) [inline src/middleware/jira-symmetric-jwt-middleware.test.ts:136]: nit: In this case, should we also drop the value from session? E.g. let's say for some reason the value got malformed. In this case, if we are not removing it from the session, the user will be stuck...

File: src/middleware/jira-symmetric-jwt-middleware.ts
Comment (bgvozdev) [inline src/middleware/jira-symmetric-jwt-middleware.ts:107]: do we have a test for this?

File: src/middleware/jira-symmetric-jwt-middleware.ts
Comment (bgvozdev) [inline src/middleware/jira-symmetric-jwt-middleware.ts:114]: do we have a test for this?

File: src/middleware/jira-symmetric-jwt-middleware.ts
Comment (bgvozdev) [inline src/middleware/jira-symmetric-jwt-middleware.ts:117]: do we have a test for this, when it is not valid?

File: src/middleware/jira-symmetric-jwt-middleware.ts
Comment (bgvozdev) [inline src/middleware/jira-symmetric-jwt-middleware.ts:36]: do we have a test for at least one of the URL from the list, to make sure we pick the token type correctly?

File: src/routes/github/github-oauth-router.ts
Comment (bgvozdev) [reply src/routes/github/github-oauth-router.ts:23]: Not sure I follow.... Has this been tested in a test?

File: src/routes/github/github-oauth-router.ts
Comment (bgvozdev) [inline src/routes/github/github-oauth-router.ts:24]: Made this change to be explicit what kind of URLs we support, which is only 2:
image

Given that `GitHubAuthMiddleware` middleware can be used in many places, we cannot rely on `req` anymore...

File: src/middleware/jira-symmetric-jwt-middleware.ts
Comment (Harminder84) [reply src/middleware/jira-symmetric-jwt-middleware.ts:118]: test case added

File: src/middleware/jira-symmetric-jwt-middleware.ts
Comment (Harminder84) [reply src/middleware/jira-symmetric-jwt-middleware.ts:121]: test case added

File: src/middleware/jira-symmetric-jwt-middleware.ts
Comment (Harminder84) [reply src/middleware/jira-symmetric-jwt-middleware.ts:39]: test case added

**Review by bgvozdev — APPROVED** [review]

**Pull request merged**
